lake/branch.go

```diff
@@ -0,0 +1,441 @@
+package lake
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/brimdata/zed/expr/extent"
+	"github.com/brimdata/zed/lake/commit"
+	"github.com/brimdata/zed/lake/index"
+	"github.com/brimdata/zed/lake/journal"
+	"github.com/brimdata/zed/lake/segment"
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/pkg/storage"
+	"github.com/brimdata/zed/zio"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zng"
+	"github.com/brimdata/zed/zson"
+	"github.com/segmentio/ksuid"
+)
+
+type BranchConfig struct {
+	Name   string      `zng:"name"`
+	ID     ksuid.KSUID `zng:"id"`
+	Parent ksuid.KSUID `zng:"parent"`
+}
+
+type Branch struct {
+	BranchConfig
+	pool   *Pool
+	engine storage.Engine
+	log    *commit.Log
+	base   commit.View
+}
+
+func newBranchConfig(name string, parent ksuid.KSUID) *BranchConfig {
+	//XXX need to take parent args and check for a cycle.
+	// also check that parent journal ID exists in parent.
+	return &BranchConfig{
+		Name:   name,
+		ID:     ksuid.New(),
+		Parent: parent,
```

Comment:
I don't grasp all the implications of doing this... but: It feels limiting to have a branch point to a parent branch vs doing what git does and having a branch to start at a particular commit. If we followed the way you have it coded up here, would you not be able to delete a branch that has child branches? What would happen to the child branches if their parent was deleted?

I suppose this would require making it so each commit referenced what its parent commit is. It would also mean that you'd have to keep commits centralized in pool... Did I mention I don't follow grasp the implications of this suggestion!?

Reply:
Good point!

Reply:
Branches v2 coming soon!